### PR TITLE
HDF5 cleanups

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Minor cleanups to testing of HDF5 infrastructure
- Add stream operator for `h5::AccessType` so we can print out its state in error messages.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
IWYU is unhappy

File: src/IO/H5/AccessType.cpp [hunk 1]
```diff
@@ -0,0 +1,12 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "IO/H5/AccessType.hpp"
+
+#include <ostream>
+
+namespace h5 {
+std::ostream& operator<<(std::ostream& os, const AccessType t) noexcept {
+  return os << (t == AccessType::ReadWrite ? "ReadWrite" : "ReadOnly");
```
Comment (nilsvu) [inline src/IO/H5/AccessType.cpp]: I prefer `switch`es for enums, since it is not clear it only has two cases.

File: src/IO/H5/Dat.cpp [hunk 2]
```diff
@@ -107,6 +109,7 @@ void Dat::append_impl(const hsize_t number_of_rows,
     }
     CHECK_H5(H5Sclose(dataspace_id), "Failed to close dataspace");
     if (read_size != size_) {
+      using ::operator<<;
```
Comment (nilsvu) [inline src/IO/H5/Dat.cpp:112]: Why is this necessary?

File: tests/Unit/IO/Test_H5.cpp [hunk 3]
```diff
@@ -38,6 +40,13 @@
 // IWYU pragma: no_include <boost/multi_array/extent_gen.hpp>
 // IWYU pragma: no_include <boost/multi_array/subarray.hpp>
 
+namespace {
+void test_access_type() noexcept {
+  CHECK(std::string{MakeString{} << h5::AccessType::ReadOnly} == "ReadOnly");
```
Comment (nilsvu) [inline tests/Unit/IO/Test_H5.cpp]: Why not use `get_output`

File: src/IO/H5/Dat.cpp (hunk 2 above)
Comment (nilsdeppe) [reply src/IO/H5/Dat.cpp:112]: Because ADL doesn't find stream operators for containers if they are not in their namespace and putting stream operators (or almost anything for that matter) into the `std` namespace is undefined behavior. This pulls in all global stream operators into the current scope so that ADL can find them.

**Comment by nilsdeppe** [comment]
@nilsleiffischer thanks for the feedback! I pushed a fixup commit and answered your question. Let me know if I was unclear or there's anything else you'd like to see changed!

**Comment by nilsvu** [comment]
LGTM 👍

**Comment by nilsdeppe** [comment]
Okay, squashed :)

**Review by nilsvu — APPROVED** [review]

**Review by kidder — APPROVED** [review]
